File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Application/States/ModeleEtatCreerPortail.swift
///////////////////////////////////////////////////////////////////////////////
/// @file ModeleEtatCreerPortail.swift
/// @author Pierre To
/// @date 2017-10-18
/// @version 1
///
/// @addtogroup log3900 LOG3990
/// @{
///////////////////////////////////////////////////////////////////////////////

import GLKit
import SceneKit

///////////////////////////////////////////////////////////////////////////
/// @class ModeleEtatCreerPortail
/// @brief Cette classe représente l'état de création des portails.
///        Implémente aussi le patron Singleton.
///
/// @author Pierre To
/// @date 2017-10-18
///////////////////////////////////////////////////////////////////////////
class ModeleEtatCreerPortail: ModeleEtat {
    
    /// Instance singleton
    static let instance = ModeleEtatCreerPortail()
    
    /// Premier portail
    private var premierNoeud: NoeudPortail?
    
    /// Fonction qui initialise l'état de création des portails
    override func initialiser() {
        /// Ignorer le tap associé au bouton pour la création de portails
        FacadeModele.instance.obtenirVue().editorView.removeGestureRecognizer(FacadeModele.instance.tapGestureRecognizer!)
        
        // Déselectionner tous les noeuds
        let arbre = FacadeModele.instance.obtenirArbreRendu()
        let table = arbre.childNode(withName: arbre.NOM_TABLE, recursively: true) as! NoeudCommun
        table.deselectionnerTout()
        
        // Envoyer la commande
        FacadeModele.instance.obtenirEtatEdition().currentUserSelectedObject(uuidSelected: "", isSelected: false, deselectAll: true)
        
        // Activer la reconnaissance de tap pour la création de portails
        FacadeModele.instance.obtenirVue().editorView.addGestureRecognizer(FacadeModele.instance.tapGestureRecognizer!)
    }
    
    /// Cette fonction nettoie l'état des changements apportes. Elle annule l'action en cours lors d'un changement d'état
    override func nettoyerEtat() {
        self.annulerCreation()
    }
    
    func annulerCreation() {
        /// Ignorer le tap associé au bouton pour la création de portails
        FacadeModele.instance.obtenirVue().editorView.removeGestureRecognizer(FacadeModele.instance.tapGestureRecognizer!)
        
        if (self.premierNoeud != nil) {
            // Supprimer le noeud ajouté
            self.premierNoeud?.removeFromParentNode()
            self.premierNoeud = nil
            FacadeModele.instance.obtenirVue().editorHUDScene?.showCancelButton(activer: false)
        }
        
        // Activer la reconnaissance de tap pour la création de portails
        FacadeModele.instance.obtenirVue().editorView.addGestureRecognizer(FacadeModele.instance.tapGestureRecognizer!)
    }
    
    // Fonctions gérant les entrées de l'utilisateur
    
    /// Évènement appelé lorsque l'utilisateur tap sur l'écran
    override func tapGesture(point: CGPoint) {
        super.tapGesture(point: point)
        
        // Création du noeud
        let arbre = FacadeModele.instance.obtenirArbreRendu()
        let noeud = arbre.creerNoeud(typeNouveauNoeud: ArbreRendu.instance.NOM_PORTAIL, uuid: "") as! NoeudPortail
            
        // Déplacement du noeud
        // Transformation du point dans l'espace virtuelle
        let convertedPoint = MathHelper.GetHitTestSceneViewCoordinates(point: self.position)
        
        if convertedPoint != nil {
            let position = GLKVector3.init(v: ((convertedPoint?.x)!, (convertedPoint?.y)!, (convertedPoint?.z)!))
            noeud.assignerPositionRelative(positionRelative: position)
        
            // Ajout du noeud à l'arbre de rendu
            let table = arbre.childNode(withName: arbre.NOM_TABLE, recursively: true) as! NoeudTable
            table.addChildNode(noeud)
            
            // Etat de l'operation
            if (self.premierNoeud == nil) {
                self.premierNoeud = noeud
                self.premierNoeud?.appliquerMaterielSelection(activer: true)
                FacadeModele.instance.obtenirVue().editorHUDScene?.showCancelButton(activer: true)
                
                // Jouer le son
                AudioService.instance.playSound(soundName: EDITION_SOUND.OBJECT1.rawValue)
            }
            else {
                // Verification que les noeuds soient sur la table
                // TODO : implémenter VisiteurSurTable
                if (self.noeudsSurLaTable()) {
                    // Link portal together
                    noeud.assignerOppose(portail: self.premierNoeud!);
                    self.premierNoeud?.assignerOppose(portail: noeud);
                    self.premierNoeud?.appliquerMaterielSelection(activer: false)
                    
                    // Envoyer la commande
                    FacadeModele.instance.obtenirEtatEdition().currentUserCreatedPortal(
                        startUuid: (self.premierNoeud?.obtenirUUID())!,
                        startPos: (self.premierNoeud?.position)!, startRotation: (self.premierNoeud?.rotation.w)!, startScale: (self.premierNoeud?.scale)!,
                        endUuid: noeud.obtenirUUID(),
                        endPos: noeud.position, endRotation: noeud.rotation.w, endScale: noeud.scale)
                    
                    self.premierNoeud = nil;
                    FacadeModele.instance.obtenirVue().editorHUDScene?.showCancelButton(activer: false)
                    
                    // Jouer le son
                    AudioService.instance.playSound(soundName: EDITION_SOUND.OBJECT2.rawValue)
                    
                    FacadeModele.instance.sauvegarderCarte(map: EditorViewController.instance.currentMap!)
                }
                else { // Annulation de la commande
                    self.annulerCreation()
                    noeud.removeFromParentNode()
                }
            }
        }
        else {
            // Afficher un message d'erreur
            FacadeModele.instance.obtenirVue().editorNotificationScene?.showErrorOutOfBoundMessage(activer: true)
        }
    }
    
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////

File: Sources/DLL/Application/States/ModeleEtatCreerPortail.cpp
///////////////////////////////////////////////////////////////////////////////
/// @file ModeleEtatCreerPortail.cpp
/// @author Nam Lesage
/// @date 2016-09-20
/// @version 0.1
///
/// @addtogroup inf2990 INF2990
/// @{
///////////////////////////////////////////////////////////////////////////////

#include <iostream>

#include "ModeleEtatCreerPortail.h"

#include "Vue.h"
#include "FacadeModele.h"
#include "ArbreRenduINF2990.h"

#include <vector>
#include "ModeleEtatJeu.h"
/// Pointeur vers l'instance unique de la classe.
ModeleEtatCreerPortail* ModeleEtatCreerPortail::instance_{ nullptr };


////////////////////////////////////////////////////////////////////////
///
/// @fn ModeleEtatCreerPortail* ModeleEtatCreerPortail::obtenirInstance()
///
/// Cette fonction retourne un pointeur vers l'instance unique de la
/// classe.  Si cette instance n'a pas été créée, elle la crée.  Cette
/// création n'est toutefois pas nécessairement "thread-safe", car
/// aucun verrou n'est pris entre le test pour savoir si l'instance
/// existe et le moment de sa création.
///
/// @return Un pointeur vers l'instance unique de cette classe.
///
////////////////////////////////////////////////////////////////////////
ModeleEtatCreerPortail* ModeleEtatCreerPortail::obtenirInstance() {
	if (instance_ == nullptr)
		instance_ = new ModeleEtatCreerPortail();

	FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->deselectionnerTout();

	return instance_;
}


////////////////////////////////////////////////////////////////////////
///
/// @fn void ModeleEtatCreerPortail::libererInstance()
///
/// Cette fonction libère l'instance unique de cette classe.
///
/// @return Aucune.
///
////////////////////////////////////////////////////////////////////////
void ModeleEtatCreerPortail::libererInstance() {
	delete instance_;
	instance_ = nullptr;
}


////////////////////////////////////////////////////////////////////////
///
/// @fn ModeleEtatCreerPortail::ModeleEtatCreerPortail()
///
/// Constructeur
///
/// @return Aucune.
///
////////////////////////////////////////////////////////////////////////
ModeleEtatCreerPortail::ModeleEtatCreerPortail()
	: ModeleEtat()
{
}


////////////////////////////////////////////////////////////////////////
///
/// @fn ModeleEtatCreerPortail::~ModeleEtatCreerPortail()
///
/// Destructeur
///
/// @return Aucune.
///
////////////////////////////////////////////////////////////////////////
ModeleEtatCreerPortail::~ModeleEtatCreerPortail() {
	if (instance_ != nullptr) {
		libererInstance();
	}
}


////////////////////////////////////////////////////////////////////////
///
/// @fn void ModeleEtatCreerPortail::mouseUpL()
///
/// Évènement appelé lorsque le bouton gauche de la souris est levé.
/// Cette fonction crée les noeuds.
///
/// @return Aucune.
///
////////////////////////////////////////////////////////////////////////
void ModeleEtatCreerPortail::mouseUpL() {
	if (mouseDownL_ && isAClick() && mouseOverTable()) {
		// Création du noeud
		NoeudPortail* noeud = (NoeudPortail*)FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->creerNoeud(ArbreRenduINF2990::NOM_PORTAIL);

		// Déplacement du noeud
		// Transformation du point dans l'esapce virtuelle
		glm::dvec3 point;
		obtenirVue()->convertirClotureAVirtuelle(mousePosX_, mousePosY_, point);

		noeud->assignerPositionRelative(point);

		// Ajout du noeud à l'arbre de rendu
		FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->ajouter(noeud);

		// Etat de l'operation	
		if (premierNoeud_ == nullptr) {
			premierNoeud_ = noeud;
			premierNoeud_->effetFantome(true);
		}
		else {
			// Verification que les noeuds soient sur la table
			if (noeudsSurLaTable()) {
				// Link portal together
				noeud->assignerOppose(premierNoeud_);
				premierNoeud_->assignerOppose(noeud);
				premierNoeud_->effetFantome(false);
				premierNoeud_ = nullptr;
				if(ModeleEtatJeu::obtenirInstance()->currentOnlineClientType() == ModeleEtatJeu::ONLINE_EDITION)
				{
					portalCreationCallback_(premierNoeud_->getUUID(),glm::value_ptr(premierNoeud_->obtenirPositionRelative()), noeud->getUUID(), glm::value_ptr(noeud->obtenirPositionRelative()));
				}
			}
			else { // Annulation de la commande
				escape();
				FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->effacer(noeud);
			}
		}
	}

	mouseDownL_ = false;
}


////////////////////////////////////////////////////////////////////////
///
/// @fn void ModeleEtatCreerPortail::escape()
///
/// Cette fonction annule l'action
///
/// @return Aucune.
///
////////////////////////////////////////////////////////////////////////
void ModeleEtatCreerPortail::escape()  {
	if (premierNoeud_ != nullptr) {
		// Supprimer le noeud ajouté
		 FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->effacer(premierNoeud_);
		 premierNoeud_->effetFantome(false);
		 premierNoeud_ = nullptr;
	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn void ModeleEtatCreerPortail::nettoyerEtat()
///
/// Cette fonction nettoie l'etat. Plus précisement, elle annule
/// l'action en cours lors d'un changement d'état.
///
/// @return Aucune.
///
////////////////////////////////////////////////////////////////////////
void ModeleEtatCreerPortail::nettoyerEtat()
{
	escape();
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////